### PR TITLE
[Feature][Hunyuan image 3.0] AR + DIT with kv reuse.

### PR DESCRIPTION
**Pull request by Bounty-hunter**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Importance: It depend on https://github.com/vllm-project/vllm-omni/pull/3107

#### 1. what can be reuse in dit stage
the prompt format in ar and dit are:
image

**Positive prompt**: reuse until the end of the CoT.
**Negtive prompt**: reuse before uncond tokens (cfg_token).

#### 2. diffusion execution flow
(1) caculate **reuse len** for pos/neg prompt according to special token: </think> </recaption> <cfg>

(2) invoke forward_call to compute the **cfg text part** for negtive prompt

(3) **update the denoise input** (we only need to compute the gen_image part)

(4) The subsequent steps remain the same.

#### 3. additional change
Refactor the ·ImageKVCacheManager` to make its logic clearer.

## Test Plan
（1）Unitest
（2）Accuracy test
（3）Performance test
```
We use CLIP to estimate the relevance between the generated CoT and the image. (higher is better)
use PSNR to evaluate the end-to-end similarity. (higher is better)
```
（4）Since we refactor the ImageKvManager, so we need to test sp, cfg parallel, tp.

## Test Result
example run:
```
python3 ../vllm-omni/examples/offline_inference/hunyuan_image3/end2end.py \
--model tencent/HunyuanImage-3.0-Instruct \
--modality img2img \
--bot-task "it2i_recaption" \
--stage-configs-path ../vllm-omni/vllm_omni/model_executor/stage_configs/hunyuan_image3_it2i_kv_reuse.yaml \
--image-path ./input_0_0.png \
--prompts "新年宠物海报，Q版圆润的可爱标题“新年快乐汪”，副标题“HAPPY NEW YEAR”。 鱼眼镜头，背景是房间门口，近景，上传的主体歪头笑，围着红色围巾，戴着红色毛线帽，高清，绒毛细节，面部特写。 宝丽莱相纸，超 surrealism，写实主义，胶片摄影，打印颗粒感肌理。肌理，超写实，复古感。" \
--seed 42 \
--steps 50 \
--guidance-scale 0.0 \
--log-stats \
--output ./results \
> ./run_reuse.log 2>&1 &
```

### Unitest
image

### Accuracy test
| case | clip score |
|-----|-----|
|   kv-reuse  | 30.1322 |
|   non-reuse 1  | 29.5113  |
|   non-reuse 2  | 30.3458 |

clip score reauslt means that the kv correct reused, and not lose the prompt feature.

| case | clip score | PSNR | SSIM |
|-----|-----|-----|-----|
|   kv-reuse vs non-reuse  | 30.1322 | 13.6 | 0.834585 |
|   non-reuse 1 vs non-reuse 2  | 29.5113  | 12.37  | 0.876087  |

It can be observed that even when executing KV no-reuse twice consecutively, the PSNR is still not very high. This is because the generated CoTs differ between runs. The differences in the CoT are caused by accumulated errors.

#### input image 735 x 1104, output image 735 x 1104
| case | end2end time | dit time | cot | image |
|-----|-----|-----|-----|-----|
|   non-reuse  | 61.88s  | 49.9s | image  | output_0_0 |
|   reuse  | 62.72s  | 49.5s | image | reuse |

#### more input len
**Theoretically, longer input sequences would yield better performance gains. However, there is currently a bug with multi-image inputs, so we will evaluate this in future work.**

Based on the actual scenario, we change `` with:
```
import random

subjects = [
    "a brown and white dog",
    "a young woman",
    "an astronaut",
    "a futuristic robot",
    "a black cat",
]

actions = [
    "running through a grassy field",
    "standing beside a river",
    "walking in the rain",
    "looking at the sunset",
    "riding a bicycle",
]

styles = [
    "cinematic lighting",
    "ultra realistic",
    "highly detailed",
    "8k resolution",
    "professional photography",
]

backgrounds = [
    "with mountains in the background",
    "inside a futuristic city",
    "under a cloudy sky",
    "surrounded by flowers",
    "in a fantasy world",
]

sentences = []

for _ in range(700):
    sentence = (
        f"{random.choice(subjects)} "
        f"{random.choice(actions)}, "
        f"{random.choice(styles)}, "
        f"{random.choice(backgrounds)}."
    )
    sentences.append(sentence)

PROMPT = " ".join(sentences)

```

```
NUM_INFERENCE_STEPS = 8
```

| case | Dit time | kv_reuse len| improve|
|-----|-----|-----|-----|
|   non-reuse  |  7.3s   |-|-|
|  reuse   |  6.6s |13126|10%|

### Performance Analyze
Using a single image input and one inference step as a case study to analyze why no performance improvement is achieved.
| case | cfg | seq len | kv reuse len |denoising loop(step = 1) | wait kv |  uncond cfg prefill  | Time Reduction |
|-----|-----|-----|----------|-----|-----|-----|-----|
|   non-reuse  |  False   |  10471   | 0 |   **1.3063s**  |  0   |   0  |  -  |
|  reuse   |  False   |   10754  | 6793 | 0.72s   |  0.1s   |  0   |  **0.5s**  |
|   non-reuse  |  True   |  10471   | 0 |  2.31s  |  0s   |   0.0s  |  -  |
|  reuse   |  True   |  10749   | 6788 | 1.12s  |  0.23s   |   0.27s  |  **0.69s** |

The first inference step is too short in duration. Theoretically, increasing the input length while reducing the number of inference steps would lead to better performance gains.

#### test for imagekvmanager (t2i)
| case | check | 
|-----|-----|
|   tp4 |  √ |
|  tp2sp2|  √  |
|   tp2cfgp2  |  √  |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by natureofnature** [comment]
@codex review

**Comment by hsliuustc0106** [comment]
Ready for full review when [WIP] status is removed. Preliminary scan available on request.

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py [hunk 1]
```diff
@@ -2827,12 +3004,15 @@ def __call__(
                 # Scheduler step (all ranks compute locally in CFG parallel)
                 latents = self.scheduler.step(pred, t, latents, **_scheduler_step_extra_kwargs, return_dict=False)[0]
                 if i != len(timesteps) - 1 and should_compute:
+                    offset = model_kwargs.get("ar_kv_reuse_offset", 0)
```
[the remark this answers is not in the document]
Comment (Bounty-hunter) [reply vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py:3017]: The ar_kv_reuse_offset is only needed on the first step when position_ids shrinks from prefill to decode length, triggering the gather branch. On subsequent steps the shapes match, so that branch is never entered and the missing key has no effect.

File: vllm_omni/model_executor/stage_input_processors/hunyuan_image3.py [hunk 2]
```diff
@@ -92,7 +94,8 @@ def ar2diffusion(
             "width": width,
             "extra": {
                 "ar_token_ids": token_tensor,
-                "ar_generated_text": generated_text,
+                "ar_generated_text": "<think>" + generated_text,
```
Comment (gcanlin) [inline vllm_omni/model_executor/stage_input_processors/hunyuan_image3.py]: Doesn't AR output `<think>`? Should we add a check here?

Comment (Bounty-hunter) [reply vllm_omni/model_executor/stage_input_processors/hunyuan_image3.py]: Yes—on the AR side, <think> is automatically added to the input, but it is not included in the output. DIT needs a flag to identify the think segment. I’ll add a `_normalize_cot_text` function in DIT to keep this logic clean and consistent.

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py [hunk 3]
```diff
@@ -2605,6 +2602,179 @@ def _split_model_kwargs_for_cfg_parallel(model_kwargs: dict[str, Any], batch_siz
                 k: v[s.start : s.stop] if isinstance(v, list) else v[s] for k, v in model_kwargs["vit_kwargs"].items()
             }
 
+    # ==========================================================
+    # Positive/ Negative Reuse Length
+    # ==========================================================
+    def _get_kv_reuse_len(self, model_kwargs, batch_size=1):
+        tokenizer_output = model_kwargs["tokenizer_output"]
+        # from positive prompt
+        pos_reuse_kv_len = tokenizer_output.think_recaption_end_pos[0][0] + 1
```
Comment (gcanlin) [inline vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: `think_recaption_end_pos` is `list[int | None]`. How can it get the [0][0] element? It seems that will crash here.

```Python
think_recaption_end_pos: list[int | None] | None = None
```

File: vllm_omni/model_executor/stage_input_processors/hunyuan_image3.py [hunk 4]
```diff
@@ -82,7 +84,8 @@ def ar2diffusion(
...
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_input_processors/hunyuan_image3.py]: `ar_full_token_ids` is added to the `extra` dict here but no consumer appears in this PR's diff. Is this consumed by code outside the diff (e.g., `prepare_model_inputs`), or is it dead code?

Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_input_processors/hunyuan_image3.py]: Hardcoded `"<think>"` prefix — this assumes the model always uses `<think>` tokens. If the chat template changes or this code is reused for a different model, the token position tracking will silently break. Consider deriving the prefix from the tokenizer's special token set instead.

**Review by hsliuustc0106 — COMMENTED** [review]
## Review Summary

**Validated:** CI green (DCO, pre-commit, build 3.11/3.12). Dependency #3107 is merged. The `ImageKVCacheManager` refactor is a genuine cleanup — the old split between `_save_image_kv_caches`/`_sp_save_prompt_kv_caches`/`_update_image_kv_caches`/`_sp_get_prompt_kv_caches` is now unified into `_cache_prompt_kv` and `_reuse_prompt_kv` with cleaner branching. New test coverage is substantial (1147 lines of unit tests across two test files, covering all three KV reuse paths plus cross-request isolation).

**Issues to address:**

1. **No test results or performance benchmarks.** The PR description has an empty "Test Result" section. This feature claims KV reuse as a performance optimization — there should be before/after timing (ms per denoising step) and memory numbers. Without benchmark data, there's no way to validate the efficiency claim. Please run the unit tests and paste output, and provide a before/after comparison for a representative generation.

2. **>1000 LOC — recommend L3 tests.** Per the [test guide](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/test_guide/#l3-level--l4-level), 1531 lines changed exceeds the L3 threshold. Could you run L3 integration tests locally and paste results here?

3. **`ar_full_token_ids` appears unused in this diff** (see inline). It's added to the `extra` dict in `ar2diffusion` but no consumer exists in the changed files. Is it needed elsewhere, or is this dead code?

4. **Hardcoded `"<think>"` string** in `hunyuan_image3.py:1953` — this assumes the model's template always uses `<think>` tags. If the template changes or a different model reuses this code, this will silently break token position tracking. Consider making this configurable or deriving it from the tokenizer.

---

*Reviewed by Claude Code with deepseek-v4-pro*

**Comment by Bounty-hunter** [comment]
Accuracy and performance tests will be added soon.

File: vllm_omni/model_executor/stage_input_processors/hunyuan_image3.py (hunk 4 above)
Comment (Bounty-hunter) [reply vllm_omni/model_executor/stage_input_processors/hunyuan_image3.py]: solved

Comment (Bounty-hunter) [reply vllm_omni/model_executor/stage_input_processors/hunyuan_image3.py]: It was temporarily used for debugging and has already been removed.

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py (hunk 3 above)
Comment (Bounty-hunter) [reply vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: It type is list[int | None] | list[list[int | None]] |  None = None , and hear is list[list[]]

**Comment by Bounty-hunter** [comment]
have added accuracy/performance test, PTAL again @gcanlin @princepride

**Comment by gcanlin** [comment]
> The first inference step is too short in duration. Theoretically, increasing the input length while reducing the number of inference steps would lead to better performance gains.

Could you provide one example that can gain the better performance?

**Comment by Bounty-hunter** [comment]
> > The first inference step is too short in duration. Theoretically, increasing the input length while reducing the number of inference steps would lead to better performance gains.
> 
> Could you provide one example that can gain the better performance?

have add in `more input len` session, it better performance with more input len

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py [hunk 5]
```diff
@@ -2602,6 +2600,189 @@ def _split_model_kwargs_for_cfg_parallel(model_kwargs: dict[str, Any], batch_siz
...
+            model_kwargs=model_kwargs,
+            batch_size=batch_size,
+            negative_reuse_len=negative_reuse_len,
+            positive_reuse_len=positive_reuse_len,
+            cfg_parallel_ready=cfg_parallel_ready,
+        )
+
+        import time
+
+        t0 = time.perf_counter()
```
Comment (gcanlin) [inline vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: I think it's not a good approach to add time counter in modeling. Maybe move it into debug block.

File: vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py [hunk 6]
```diff
@@ -1557,6 +1557,7 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
 
         self._replace_rotary_embeddings()
         self._patch_moe_blocks()
+        self.eval()
```
Comment (princepride) [inline vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py]: I think we don't need it

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py [hunk 7]
```diff
@@ -1295,6 +1363,7 @@ def forward(
             extra = p.get("extra", {}) if isinstance(p, dict) else {}
             cot_text_list.append(extra.get("ar_generated_text") or None)
         cot_text = cot_text_list if any(t is not None for t in cot_text_list) else None
+        cot_text = [self._normalize_cot_text(t) for t in cot_text] if cot_text is not None else None
```
Comment (princepride) [inline vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py]: Can we combine this two lines?

File: examples/offline_inference/hunyuan_image3/end2end.py [hunk 8]
```diff
@@ -228,7 +229,12 @@ def main():
     print(f"{'=' * 60}\n")
 
     # Generate
+    import time
+
+    t0 = time.perf_counter()
```
Comment (princepride) [inline examples/offline_inference/hunyuan_image3/end2end.py]: Maybe we can remove these.

**Review by princepride — CHANGES_REQUESTED** [review]
Just a little suggestion

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py (hunk 5 above)
Comment (Bounty-hunter) [reply vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: fix it by use `logger.isEnabledFor(logging.DEBUG)` to control.

File: examples/offline_inference/hunyuan_image3/end2end.py (hunk 8 above)
Comment (Bounty-hunter) [reply examples/offline_inference/hunyuan_image3/end2end.py]: fixed

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py (hunk 7 above)
Comment (Bounty-hunter) [reply vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py]: fixed

File: vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py (hunk 6 above)
Comment (Bounty-hunter) [reply vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py]: you are right, `"attention_dropout": 0.0,` in config

File: vllm_omni/model_executor/stage_configs/hunyuan_image3_it2i_kv_reuse.yaml [hunk 9]
```diff
@@ -0,0 +1,89 @@
+stage_args:
```
Comment (gcanlin) [inline vllm_omni/model_executor/stage_configs/hunyuan_image3_it2i_kv_reuse.yaml:1]: Do we plan to wait deploy config or merge this first? @Fishermanykx

File: vllm_omni/model_executor/stage_configs/hunyuan_image3_it2i_kv_reuse.yaml [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    stage_type: llm
+    runtime:
+      process: true
+      devices: "0,1"
+      max_batch_size: 1
+      requires_multimodal_data: true  # AR needs the original image
+    engine_args:
+      model_stage: AR
+      model_arch: HunyuanImage3ForCausalMM
+      worker_cls: vllm_omni.worker.gpu_ar_worker.GPUARWorker
```
Comment (gcanlin) [inline vllm_omni/model_executor/stage_configs/hunyuan_image3_it2i_kv_reuse.yaml:13]: ~Remove `worker_cls`~

**Review by gcanlin — APPROVED** [review]
LGTM, just need to confirm the config.

**Review by princepride — APPROVED** [review]

**Comment by gcanlin** [comment]
@Fishermanykx Could you please run this test locally? From the current config, it seems that can't run on NPU. If need much effort, leaving to next PR is fine.

**Comment by Fishermanykx** [comment]
> @Fishermanykx Could you please run this test locally? From the current config, it seems that can't run on NPU. If need much effort, leaving to next PR is fine.

Tested. It only need minor adjustment in the yaml file to run on NPU. I'll adjust deploy yaml according to this pr
